test/e2e/rebuild/basic_rebuild_test.go

```diff
@@ -0,0 +1,99 @@
+
+import (
+	"e2e-basic/common"
+	"testing"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+	"time"
+)
+
+var (
+	pvcName      = "rebuild-test-pvc"
+	storageClass = "mayastor-nvmf"
+)
+
+func basicRebuildTest() {
+	pvc, err := common.GetPVC(pvcName)
+	Expect(err).To(BeNil())
+	Expect(pvc).ToNot(BeNil())
+
+	uuid := string(pvc.ObjectMeta.UID)
+	repl, err := common.GetNumReplicas(uuid)
+	Expect(err).To(BeNil())
+	Expect(repl).Should(BeEquivalentTo(1))
+
+	// Add another child which should kick off a rebuild.
+	common.UpdateNumReplicas(uuid, 2)
+	repl, err = common.GetNumReplicas(uuid)
+	Expect(err).To(BeNil())
+	Expect(repl).Should(BeEquivalentTo(2))
+
+	timeout := "90s"
+	pollPeriod := "1s"
+
+	// Wait for the added child to show up.
+	Eventually(func() int { return common.GetNumChildren(uuid) }, timeout, pollPeriod).Should(BeEquivalentTo(2))
+
+	getChildrenFunc := func(uuid string) []common.NexusChild {
+		children, err := common.GetChildren(uuid)
+		if err != nil {
+			panic("Failed to get children")
+		}
+		Expect(len(children)).Should(BeEquivalentTo(2))
+		return children
+	}
+
+	// Check the added child and nexus are both degraded.
+	Eventually(func() string { return getChildrenFunc(uuid)[1].State }, timeout, pollPeriod).Should(BeEquivalentTo("CHILD_DEGRADED"))
+	Eventually(func() (string, error) { return common.GetNexusState(uuid) }, timeout, pollPeriod).Should(BeEquivalentTo("NEXUS_DEGRADED"))
+
+	// Check everything eventually goes healthy following a rebuild.
+	Eventually(func() string { return getChildrenFunc(uuid)[0].State }, timeout, pollPeriod).Should(BeEquivalentTo("CHILD_ONLINE"))
+	Eventually(func() string { return getChildrenFunc(uuid)[1].State }, timeout, pollPeriod).Should(BeEquivalentTo("CHILD_ONLINE"))
+	Eventually(func() (string, error) { return common.GetNexusState(uuid) }, timeout, pollPeriod).Should(BeEquivalentTo("NEXUS_ONLINE"))
+}
+
+func TestRebuild(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Rebuild Test Suite")
+}
+
+var _ = Describe("Mayastor rebuild test", func() {
+	It("should run a rebuild job to completion", func() {
+		basicRebuildTest()
+	})
+})
+
+var _ = BeforeSuite(func(done Done) {
+	logf.SetLogger(zap.LoggerTo(GinkgoWriter, true))
+	common.SetupTestEnv()
+	common.MkPVC(pvcName, storageClass)
+	CreateDummyPod()
+	close(done)
+}, 60)
+
+var _ = AfterSuite(func() {
+	By("tearing down the test environment")
+	DestroyDummyPod()
+	common.RmPVC(pvcName, storageClass)
+	common.TeardownTestEnv()
+})
+
+const ApplicationPod = "fio.yaml"
+
+// CreateDummyPod deploys a pod which mounts the PVC but doesn't do anything with it.
+// This is to allow a nexus entry to be created in the MSV (because a nexus is only created on volume publish).
+func CreateDummyPod() {
+	common.ApplyDeployYaml(ApplicationPod)
+	// Allow some time for the PVC to be mounted.
+	time.Sleep(3 * time.Second)
```

Comment:
Is there a way how we could test that the volume has been mounted without using fixed time intervals? I think we should have a utility function in common library that executes a callback with exponential backoff until successful. Just like we have in mocha tests. No idea how difficult it is to implement in golang. But I suppose it would have many use cases in other tests as well.

Reply:
@jkryl  I think the idiom is `Eventually` for example in `pvc_stress` test we have
```
	// Wait for the PVC to be bound.
	Eventually(func() coreV1.PersistentVolumeClaimPhase {
		return Cmn.GetPvcStatusPhase(volName)
	},
		defTimeoutSecs, // timeout
		"1s",           // polling interval
	).Should(Equal(coreV1.ClaimBound))
```

Reply:
The code has been refactored a bit, but I have replaced the sleep with the check that @blaisedias suggested.

The sleep was required to ensure that the volume was published before adding a child. This is needed to address the race condition detailed [here](https://github.com/openebs/Mayastor/issues/590). However, after speaking with @jkryl I have included an additional function to check if a volume has been published. This checks for the existence of the "Target Nodes" field in the MSV. If this field exists, the volume is published.